library/src/main/scala/scala/scalajs/concurrent/QueueExecutionContext.scala

```diff
@@ -1,22 +1,55 @@
 package scala.scalajs.concurrent
 
 import scala.concurrent.ExecutionContextExecutor
-import scalajs.js
 
-private[concurrent] object QueueExecutionContext
-    extends ExecutionContextExecutor {
+import scala.scalajs.js
+import scala.scalajs.js.|
 
-  def execute(runnable: Runnable): Unit = {
-    js.timers.setTimeout(0) {
-      try {
-        runnable.run()
-      } catch {
-        case t: Throwable => reportFailure(t)
-      }
+object QueueExecutionContext {
+  def timeouts(): ExecutionContextExecutor =
+    new TimeoutsExecutionContext
+
+  def promises(): ExecutionContextExecutor =
+    new PromisesExecutionContext
+
+  def apply(): ExecutionContextExecutor =
+    if (js.isUndefined(js.Dynamic.global.Promise)) timeouts()
+    else promises()
+
+  private final class TimeoutsExecutionContext extends ExecutionContextExecutor {
+    def execute(runnable: Runnable): Unit = {
+      js.Dynamic.global.setTimeout({ () =>
```

Comment:
I know. But as it is now, there is no way to have a "stable" execution context. I.e. one that doesn't depend on global state. Usually it is not good if stuff depends on global state. Since we don't really have a choice here, we could go with the nicer: "We capture state when you call xyz", rather than: "State just updates under the hood, tough luck, pal"...

OTOH, I don't know if there are VMs that do not like if one captures `setTimeout` (also, it is not evident whether we must capture global to have a proper `this` argument. And if we do, the whole thing is kind of bogus.).